### ISSUE TITLE
About Discriminative Model Loss Function（Bug？？？）

### ISSUE THREAD

**Issue by wenfeixiang1991** [HazyResearch/snorkel#877]
**There are some knowledge about discriminative model loss function in Snorkel Project.**

1. [Snorkel: Rapid Training Data Creation with Weak Supervision (VLDB 2018)](https://arxiv.org/abs/1711.10160).It mainly saids two points about loss function. 
The first one just give a discretionally formula expression which I could know that 

> we train a discriminative model hθ on our probabilistic labels Y ̃ by minimizing a noise-aware variant of the loss l(hθ(xi),y), i.e., the expected loss with respect to Y ̃

   The second one give a words description.

> To best take advantage of this, Snorkel creates probabilistic training labels for any discrim- inative model with a standard loss function.
2. Snorkel code [disc_learning.py](https://github.com/HazyResearch/snorkel/blob/0b76242bb76098967b736b15aa699b32979bdc6b/snorkel/learning/disc_learning.py#L71) just uses `tf.nn.sigmoid_cross_entropy_with_logits` and I don't see any changes of loss function.

**So, here are my questions** 

1. What does 'standard loss function' in [Snorkel: Rapid Training Data Creation with Weak Supervision (VLDB 2018)](https://arxiv.org/abs/1711.10160) mean?
2. Is that snorkel code [disc_learning.py](https://github.com/HazyResearch/snorkel/blob/0b76242bb76098967b736b15aa699b32979bdc6b/snorkel/learning/disc_learning.py#L71) `tf.nn.sigmoid_cross_entropy_with_logits` right? Because in my point of view, cross entropy function is `Loss = -1/n * sum(y~*ln(y) + (1-y~)*ln(1-y))` which is only effective when y~ is 0 or 1, if y~ is some probabilistic number between 0 and 1, I can't see it works directly. But `1/2*(y - y~)^2` does make sense when y~ is probabilistic label.
3. If I need change loss function to match discriminative model in snorkel, how would I do that? 
4. The reason I post this issue is that the loss is not decrease when I use snorkel code re_rnn.py, which is #870 . Then I check every mistakes I probably make, nothing found but one point I notice that most my input_y is around 0.5. So I thought this could cause loss not decrease, but in snorkel theory, y always approach y~, and I didn't see that could be happen when the code is `tf.nn.sigmoid_cross_entropy_with_logits`
5. By the way, there is another bug in [rnn_base.py](https://github.com/HazyResearch/snorkel/blob/0b76242bb76098967b736b15aa699b32979bdc6b/snorkel/learning/disc_models/rnn/rnn_base.py#L116), it looks like `potentials_dropout` is useless, which mean `dropout` is useless.

**Thank you very much!**

.

**Comment by wenfeixiang1991** [comment]
It works when I changed the label，because lots of label‘s probabilities are 0.5 and I don't think the default loss function in tensorflow is right in this circumstances，but snorkel code just uses sigmoid_corss_entropy_with_logits and I am confused!

**Comment by stephenbach** [comment]
Hi!

Thanks for your interest in Snorkel!

By 'standard,' we're emphasizing that the framework is modular and any of a number of common loss functions. What's neat is that if your loss function is the negative log likelihood, then the expected loss is just the cross entropy loss with probabilities in [0,1] for \tilde{y}. So that code is doing the right thing!
